Zend/zend_gc.c

```diff
@@ -569,128 +762,34 @@ static void gc_mark_grey(zend_refcounted *ref)
 			zv = Z_INDIRECT_P(zv);
 		}
 		ref = Z_COUNTED_P(zv);
-		GC_REFCOUNT(ref)--;
-		goto tail_call;
-	}
-}
-
-static void gc_mark_roots(void)
-{
-	gc_root_buffer *current = GC_G(roots).next;
-
-	while (current != &GC_G(roots)) {
-		if (GC_REF_GET_COLOR(current->ref) == GC_PURPLE) {
-			gc_mark_grey(current->ref);
-		}
-		current = current->next;
-	}
-}
-
-static void gc_scan(zend_refcounted *ref)
-{
-    HashTable *ht;
-	Bucket *p, *end;
-	zval *zv;
-
-tail_call:
-	if (GC_REF_GET_COLOR(ref) == GC_GREY) {
```

Comment:
As white and black marking happens simultaneously, I think we should check here that color of the node is white as if it's black  the children are already marked black as well.

Reply:
done